File: solutions/003-longest-substring-without-repeating-characters/solution.go
package main

import "fmt"

func lengthOfLongestSubstring(s string) int {
	n := len(s)
	if n == 0 {
		return 0
	}
	maxLen := 1
	left := 0
	for right := 1; right < n; right++ {
		// 检查 s[right] 是否在 [left, right-1] 窗口内重复
		for i := left; i < right; i++ {
			if s[i] == s[right] {
				left = i + 1 // 重复则左边界移到重复字符的下一个
				break
			}
		}
		// 当前窗口长度 right-left+1，取最大值
		if curLen := right - left + 1; curLen > maxLen {
			maxLen = curLen
		}
	}
	return maxLen
}

func main() {
	tests := []struct{ s string; want int }{
		{"abcabcbb", 3}, {"bbbbb", 1}, {"pwwkew", 3},
	}
	for _, tc := range tests {
		got := lengthOfLongestSubstring(tc.s)
		status := "PASS"
		if got != tc.want {
			status = "FAIL"
		}
		fmt.Printf("%s | s=%q => %d (expected %d)\n", status, tc.s, got, tc.want)
	}
}

File: solutions/003-longest-substring-without-repeating-characters/solution.py
class Solution:
    def length_of_longest_substring(self, s: str) -> int:
        n = len(s)
        if n == 0:
            return 0
        max_len = 1
        left = 0
        for right in range(1, n):
            for i in range(left, right):
                if s[i] == s[right]:
                    left = i + 1
                    break
            cur_len = right - left + 1
            if cur_len > max_len:
                max_len = cur_len
        return max_len


if __name__ == "__main__":
    tests = [("abcabcbb", 3), ("bbbbb", 1), ("pwwkew", 3)]
    sol = Solution()
    for s, expected in tests:
        got = sol.length_of_longest_substring(s)
        status = "PASS" if got == expected else "FAIL"
        print(f"{status} | s={s!r} => {got} (expected {expected})")
